kolibri/core/assets/test/api-resource.js

```diff
@@ -208,3 +248,893 @@ describe('Resource', function () {
+        this.collection.set([this.model, this.model]);
+        assert.deepEqual(this.collection._model_map, { [this.model.id]: this.model });
+      });
+    });
+  });
+});
+
+describe('Model', function () {
+  beforeEach(function () {
+    this.resource = {
+      modelUrl: () => '',
+      idKey: 'id',
+    };
+    this.data = { test: 'test', id: 'testing' };
+    this.model = new Resources.Model(this.data, this.resource);
+  });
+  afterEach(function () {
+    delete this.resource;
+    delete this.model;
+  });
+  describe('constructor set properties:', function () {
+    describe('resource property', function () {
+      it('should be the passed in resource', function () {
+        assert.equal(this.resource, this.model.resource);
+      });
+    });
+    describe('attributes property', function () {
+      it('should be the data', function () {
+        assert.deepEqual(this.model.attributes, this.data);
+      });
+    });
+    describe('synced property', function () {
+      it('should be false', function () {
+        assert.equal(this.model.synced, false);
+      });
+    });
+    describe('promises property', function () {
+      it('should be an empty array', function () {
+        assert.deepEqual(this.model.promises, []);
+      });
+    });
+  });
+  describe('constructor method', function () {
+    describe('if resource is undefined', function () {
+      it('should throw a TypeError', function () {
+        assert.throws(() => new Resources.Model(this.data), TypeError);
+      });
+    });
+    describe('if data is passed in', function () {
+      it('should call the set method once', function () {
+        const spy = sinon.spy(Resources.Model.prototype, 'set');
+        const testModel = new Resources.Model(this.data, this.resource);
+        assert.ok(testModel);
+        assert.ok(spy.calledOnce);
+        Resources.Model.prototype.set.restore();
+      });
+      it('should call the set method with the data', function () {
+        const spy = sinon.spy(Resources.Model.prototype, 'set');
+        const testModel = new Resources.Model(this.data, this.resource);
+        assert.ok(testModel);
+        assert.ok(spy.calledWithExactly(this.data));
+        Resources.Model.prototype.set.restore();
+      });
+    });
+    describe('if undefined data is passed in', function () {
+      it('should throw a TypeError', function () {
+        assert.throws(() => new Resources.Model(undefined, this.resource), TypeError);
+      });
+    });
+    describe('if null data is passed in', function () {
+      it('should throw a TypeError', function () {
+        assert.throws(() => new Resources.Model(null, this.resource), TypeError);
+      });
+    });
+    describe('if no data is passed in', function () {
+      it('should throw a TypeError', function () {
+        assert.throws(() => new Resources.Model({}, this.resource), TypeError);
+      });
+    });
+  });
+  describe('fetch method', function () {
```

Comment:
This is broadly copied and modified from the Collection `fetch` method, and the `save` and `delete` are improvisations around the `fetch` tests.